hilti/toolchain/src/compiler/global-optimizer.cc

```diff
@@ -753,6 +753,155 @@ struct ConstantFoldingVisitor : OptimizerVisitor, visitor::PreOrder<bool, Consta
+
+        // We only work on feature flags named `__feat*`.
+        if ( ! util::startsWith(id, "__feat") )
+            return;
+
+        const auto& tokens = util::split(id, "%");
+        assert(tokens.size() == 3);
+
+        // The type name is encoded in the variable name with `::` replaced by `__`.
+        const auto& typeID = ID(util::replace(tokens[1], "__", "::"));
+        const auto& feature = tokens[2];
+
+        switch ( _stage ) {
+            case Stage::COLLECT: {
+                // Record the feature as unused for the type if it was not already recorded.
+                _features[typeID].insert({feature, false});
+                break;
+            }
+
+            case Stage::TRANSFORM: {
+                const auto required = _features.at(typeID).at(feature);
+                const auto value = x.init().value().as<expression::Ctor>().ctor().as<ctor::Bool>().value();
+
+                if ( required != value ) {
+                    HILTI_DEBUG(logging::debug::GlobalOptimizer,
+                                util::fmt("disabling feature '%s' of type '%s' since it is not used", feature, typeID));
+
+                    auto new_x = declaration::GlobalVariable::setInit(x, builder::bool_(false));
+                    replaceNode(p, new_x);
+                }
+
+                break;
+            }
+        }
+    }
+
+    template<typename MemberAccess>
+    void handleMemberAccess(const MemberAccess& x, position_t p) {
+        switch ( _stage ) {
+            case Stage::COLLECT: {
+                auto type_ = x.op0().type();
+                while ( type::isReferenceType(type_) )
+                    type_ = type_.dereferencedType();
+
+                auto typeID = type_.typeID();
+                if ( ! typeID )
+                    return;
+
+                auto member = x.op1().template tryAs<expression::Member>();
+                if ( ! member )
+                    return;
+
+                auto lookup = scope::lookupID<declaration::Type>(*typeID, p, "type");
+                if ( ! lookup )
+                    return;
+
+                auto type = lookup->first->template as<declaration::Type>();
+                auto struct_ = type.type().template tryAs<type::Struct>();
+                if ( ! struct_ )
+                    return;
+
+                auto field = struct_->field(member->id());
+                if ( ! field )
+                    return;
+
+                // Compute a list of features this use does not activate.
+                // Generated feature-dependent code is always under
+                // conditionals `if (__feat%XYZ%FEATURE) ...` so get all
+                // `FEATURE` for all parents which match this pattern.
+                const std::vector<std::string> ignored_features =
+                    util::transform(util::filter(util::transform(p.path,
+                                                                 [](const visitor::Location<Node&>& n)
```

Comment:
Would factor this out into a standalone function, gets pretty deeply nested here.

Reply:
Yeah, I went over board here trying to do `filter_map`/`flatMap` 😄 . Since we would allocate for tons of intermediate `vector`s here, and I'd really prefer a `set` instead, moving this to a function shouldn't just improve readability, but also performance.